File: samples/sample_s390x.c
/* Unicorn Emulator Engine */
/* By Nguyen Anh Quynh, 2021 */

/* Sample code to demonstrate how to emulate S390X code */

#include <unicorn/unicorn.h>
#include <string.h>

// code to be emulated
#define S390X_CODE "\x18\x23" // lr %r2, %r3

// memory address where emulation starts
#define ADDRESS 0x10000

static void hook_block(uc_engine *uc, uint64_t address, uint32_t size,
                       void *user_data)
{
    printf(">>> Tracing basic block at 0x%" PRIx64 ", block size = 0x%x\n",
           address, size);
}

static void hook_code(uc_engine *uc, uint64_t address, uint32_t size,
                      void *user_data)
{
    printf(">>> Tracing instruction at 0x%" PRIx64
           ", instruction size = 0x%x\n",
           address, size);
}

static void test_s390x(void)
{
    uc_engine *uc;
    uc_hook trace1, trace2;
    uc_err err;

    uint64_t r2 = 2, r3 = 3;

    printf("Emulate S390X code\n");

    // Initialize emulator in S390X mode
    err = uc_open(UC_ARCH_S390X, UC_MODE_BIG_ENDIAN, &uc);
    if (err) {
        printf("Failed on uc_open() with error returned: %u (%s)\n", err,
               uc_strerror(err));
        return;
    }

    // map 1MB memory for this emulation
    uc_mem_map(uc, ADDRESS, 1024 * 1024, UC_PROT_ALL);

    // write machine code to be emulated to memory
    uc_mem_write(uc, ADDRESS, S390X_CODE, sizeof(S390X_CODE) - 1);

    // initialize machine registers
    uc_reg_write(uc, UC_S390X_REG_R2, &r2);
    uc_reg_write(uc, UC_S390X_REG_R3, &r3);

    // tracing all basic blocks with customized callback
    uc_hook_add(uc, &trace1, UC_HOOK_BLOCK, hook_block, NULL, 1, 0);

    // tracing all instruction
    uc_hook_add(uc, &trace2, UC_HOOK_CODE, hook_code, NULL, 1, 0);

    // emulate machine code in infinite time (last param = 0), or when
    // finishing all the code.
    err = uc_emu_start(uc, ADDRESS, ADDRESS + sizeof(S390X_CODE) - 1, 0, 0);
    if (err) {
        printf("Failed on uc_emu_start() with error returned: %u (%s)\n", err,
               uc_strerror(err));
    }

    // now print out some registers
    printf(">>> Emulation done. Below is the CPU context\n");

    uc_reg_read(uc, UC_S390X_REG_R2, &r2);
    uc_reg_read(uc, UC_S390X_REG_R3, &r3);

    printf(">>> R2 = 0x%" PRIx64 "\t\t>>> R3 = 0x%" PRIx64 "\n", r2, r3);

    uc_close(uc);
}

int main(int argc, char **argv, char **envp)
{
    test_s390x();

    return 0;
}

File: bindings/python/sample_s390x.py
#!/usr/bin/env python

from unicorn import *
from unicorn.s390x_const import *

# lr %r2, %r3
S390X_CODE = b"\x18\x23"

# memory address where emulation starts
ADDRESS    = 0x10000


# callback for tracing basic blocks
def hook_block(uc, address, size, user_data):
    print(">>> Tracing basic block at 0x%x, block size = 0x%x" %(address, size))


# callback for tracing instructions
def hook_code(uc, address, size, user_data):
    print(">>> Tracing instruction at 0x%x, instruction size = 0x%x" %(address, size))


# Test RISCV
def test_s390x():
    print("Emulate S390X code")
    try:
        # Initialize emulator in big endian mode
        mu = Uc(UC_ARCH_S390X, UC_MODE_BIG_ENDIAN)

        # map 2MB memory for this emulation
        mu.mem_map(ADDRESS, 2 * 1024 * 1024)

        # write machine code to be emulated to memory
        mu.mem_write(ADDRESS, S390X_CODE)

        # initialize machine registers
        mu.reg_write(UC_S390X_REG_R3, 0x7890)

        # tracing all basic blocks with customized callback
        mu.hook_add(UC_HOOK_BLOCK, hook_block)

        # tracing all instructions with customized callback
        mu.hook_add(UC_HOOK_CODE, hook_code)

        # emulate machine code in infinite time
        mu.emu_start(ADDRESS, ADDRESS + len(S390X_CODE))

        # now print out some registers
        print(">>> Emulation done. Below is the CPU context")

        r2 = mu.reg_read(UC_S390X_REG_R2)
        r3 = mu.reg_read(UC_S390X_REG_R3)
        print(">>> R2 = 0x%x" % r2)
        print(">>> R3 = 0x%x" % r3)

    except UcError as e:
        print("ERROR: %s" % e)


if __name__ == '__main__':
    test_s390x()

